### ISSUE TITLE
Introduce support for diffing subsets of data in generic RecyclerView adapter

### ISSUE THREAD

**Issue by BenHenning** [oppia/oppia-android#171]
The initial implementation of the generic RecyclerView BindableAdapter notifies that the whole data set is changed for each re-bind, which is sometimes wrong (the data may be the same) and highly inefficient (1 element might have changed out of hundreds). Diffing logic should be introduced to only notify that a subset of the data has changed, which also enables support for RecyclerView animations.

Later update for much more context on approaching this issue:

We've tried two attempts at fixing this:
- #1359 that was actually merged, but then reverted in #1627 due to causing #1625 and #1626. As far as I know, we never looked into the specifics of what caused the performance issues here but I have some suspicions (see below)
- #2666 there were a bunch of test failures that weren't fixed, and more manual testing was needed. It also wasn't quite implemented correctly as we want to use more granular controls than just ``notifyDataSetChanged``.

To illustrate the current state and desired outcome (for added context):
- Today, ``BindableAdapter`` has a single ``setData`` method that replaces the entire list of data: https://github.com/oppia/oppia-android/blob/4f5aec630c8b4826cd94fe63bc2ab99f0cb35ad7/app/src/main/java/org/oppia/android/app/recyclerview/BindableAdapter.kt#L42-L48 This is actually very convenient because it ties directly into databinding. However, it is very inefficient: calling [``notifyDataSetChanged``](https://developer.android.com/reference/androidx/recyclerview/widget/RecyclerView.Adapter#notifyDataSetChanged()) which is a bit of a sledgehammer solution since it basically forces the recycler view to reset its views (which causes issues like #4839 plus other annoyances like view flashing and prevents clean automatic built-in recycler view animations).
- Ideally, we would actually use the more granular adapter functions for notifying changes: ``notifyItemChanged``, ``notifyItemInserted``, and ``notifyItemRemoved`` (possibly also ``notifyItemRangeChanged``, ``notifyItemRangeInserted``, and ``notifyItemRangeRemoved`` though these are less important and probably harder to do) so that ``RecyclerView`` knows how to manage incoming data changes correctly.
- Since we can't easily add new ``set*()`` methods in ``BindableAdapter`` for this purpose, we need to rely on diffing to understand the difference between the current set of data bound to the recycler view and the new list that's incoming. ``DiffUtil`` performs this computation. It has a benchmark listed that we could try to beat, but I'm not sure that it's necessary. I think we should try to use ``DiffUtil`` first, but correctly.
- #1359 attempted to use ``DiffUtil`` by directly comparing bound elements rather than doing anything specific to the data itself. The hope was that this would be a reasonable stop-gap. It turns out it wasn't and led to worse performance issues.
- I strongly suspect that #170 will be a prerequisite here. The idea of introducing [stable IDs](https://developer.android.com/reference/androidx/recyclerview/widget/RecyclerView.Adapter#hasStableIds()) is technically more for the purpose of animating, but I think it also adds internal performance benefits for recycler view. More to the point: stable IDs makes it much easier for us to perform correct diffing when implementing ``DiffUtil.Callback``. I suspect that #1359 had performance issues because it was causing ``areContentsTheSame`` to be called far too often, and in a way that probably has roughly O(n^2) complexity.

Introducing the necessary functionality for a correct ``DiffUtil.Callback`` requires establishing a strict API requirement for data bound to ``BindableAdapter``. I think that's fine, but it will involve a potentially large scale refactor in the codebase (and isn't easy to stage). That would require changing ``BindableAdapter`` to extend a new ``BindableItemViewModel`` class instead of ``Any``, and updating all uses of ``BindableAdapter`` in the codebase to adhere to this new class (which is hopefully as simple as implementing the interface in a bunch of new places). I think this interface could look like the following (as a concept):

```kotlin
/** Represents a single item in a bindable recycler view. */
interface BindableItemViewModel<T: BindableItemViewModel<T>> {
  /**
   * The unique content ID for this item. Note that content IDs are expected to be unique
   * even for view models of the same type, and should never be reused for a new view model.
   * Note that objects used as IDs must be able to be checked correctly for equivalence using
   * ``equals``.
   */
  val contentId: Any

  /**
   * Returns whether the [other] version of this view model has different contents to this version of it. This is used for data diffing analysis and should only compare fields that are relevant to rendering the view model.
   *
   * Note that data subclasses can probably just use an equals check for this so long as all of their contents are also data classes or have correct equals overrides (and it makes sense to check all fields).
   */
  fun hasChanges(other: T): Boolean
}
```

We could, alternatively, update ``BindableAdapter``'s builders to provide per-item ID and comparison methods (similar to what we do today for typing), but that seems a bit unwieldy. It may also be worth considering refactoring those builders to move the type information here, but I don't know how involved that will be since the single vs. multi typing is fairly complex in the adapter.

If we implemented the above, the ``DiffUtil.Callback`` becomes relatively trivial:

```kotlin
class BindableAdapterDiffUtilHandler<T: BindableItemViewModel>(
  private val oldList: List<T>,
  private val newList: List<T>
): DiffUtil.Callback {
  override fun getOldListSize() = oldList.size
  override fun getNewListSize() = newList.size
  override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int) =
    oldList[oldItemPosition].contentId == newList[newItemPosition].contentId
  override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int) =
    oldList[oldItemPosition].hasChanges(newList[newItemPosition])
}
```

This could even be extended in the future to introduce sub-item diffing (e.g. for item animation) by extending the contract of ``hasChanges`` to perform a sub-item diff and then hooking it up to ``getChangePayload`` in the custom diff util callback above. I didn't include that here because of it being unclear how to generalize the item animation side of it (even we would even want to do that).

Now the main part comes: how do we implement unique content IDs and ``hasChanges`` across all of the possible view models? Well:
- ``hasChanges`` is actually straightforward. We only need to cross-compare fields that are directly relevant to internal state and/or aspects of the view model that are displayed. This is likely most or all of the data in the view model (though injectable objects should be ignored). This should never be a direct equals check except for specific kinds of data classes.
- I suspect we will need to introduce custom ID classes on a per recycler view basis--see my example below.

An example of an adapter instance-specific ID would be as follows for the exploration player (not complete, just a basic example):

```kotlin
sealed class StateItemId {
  data class Content(val contentId: String): StateItemId()
  data class Feedback(val feedbackId: String): StateItemId()
  object PreviousNavigationButton: StateItemId() // Only one of these can ever exist.
  ... // See StateItemViewModel.ViewType for other types.
}
```

(Note that subclasses are data classes and are able to be checked for equivalence).

(Note for refactoring: if we represented this sealed class as a proto ``oneof`` then we could use the corresponding proto enum generated to also replace ``StateItemViewModel.ViewType`` which would probably be a nice way to combine the two concepts together).

Finally, I want to note a few things on approaching this project:
1. We should establish the high-level patterns of the mechanisms needed _and_ implement them in a very simple adapter case in the same PR to demonstrate it working. One way to do this would be updating ``BindableAdapter`` to have a temporary new ``setData`` method that checks the actual data types for subclassing ``BindableItemViewModel`` without updating the adapter's generic type. This allows us to introduce a small runtime cost to demonstrate that the functionality works with a single adapter without needing to implement it everywhere.
2. We should ensure that the new diffing functionality is gated by a feature flag so that if it causes performance or stability regressions that these aren't seen by users until we have time to test it thoroughly.
3. We should implement support for all other adapters one-by-one (which will likely be multiple PRs since some of the adapter uses are fairly complex).
4. Once all adapters are updated to use the new view model, then we can remove the temporary function and ``setData`` method.
5. We need to manually test with the feature on to make sure it works well. After manual testing, we may do additional testing through our alpha release channel if it's warranted (but, in the least, we'll also do additional QA testing since it's likely this will cause unforeseen issues particularly in the exploration player).
6. We need to think through automated testing and ensure that it's done well both for the new adapter support (i.e. that diffing correctly notifies on a per-item basis as we expect--note that this is essentially testing ``DiffUtil`` but that's ideal since if the utility isn't doing what we want, we'll need to introduce our own diffing logic) and for each view model that uses it. The latter is actually much more complex as this new logic may be able to result in item updates being ignored. This probably involves auditing existing tests for each adapter use case to ensure that the tests correctly verify scenarios where the corresponding recycler view's data changes in a way that would trigger each of the different diff util scenarios we care about (specifically item added, item removed, and item changed).

**Comment by anandwana001** [comment]
@BenHenning Tried this issue using `AsyncListDiffer` and setting the new data inside our `setData(newDataList: List<T>)` with `differ.submitList(buildNewList(newDataList))`

Video link - https://drive.google.com/drive/folders/1HAfzcLl2BkSddAjuUuCBJAmf5RwP9Atn?usp=sharing

Before
![Before](https://user-images.githubusercontent.com/25057618/79196034-f0282b00-7e4c-11ea-8267-9f2aa57cda54.gif)

After
![After](https://user-images.githubusercontent.com/25057618/79196440-bd326700-7e4d-11ea-8f46-d19ea0cab5a7.gif)

**Comment by BenHenning** [comment]
Reopened due to revert: https://github.com/oppia/oppia-android/pull/1627.

**Comment by khushal1707** [comment]
@anandwana001 I like to take this issue, could you give me few pointers on what was the problem with implementation of DiffUtil?  Were test cases complete and passing?

**Comment by anandwana001** [comment]
@khushal1707 
Things which are failing:
1. RecyclerView loading is smooth but slow
2. Drag and Drop doesn't work correctly

These are the 2 issues we face after merging my earlier pr
1. https://github.com/oppia/oppia-android/issues/1625
2. https://github.com/oppia/oppia-android/issues/1626

**Comment by aggarwalpulkit596** [comment]
@anandwana001 didn't we solved this earlier ?

**Comment by anandwana001** [comment]
No, the linking thing in drag and drop doesn't work with the diffing.

**Comment by aggarwalpulkit596** [comment]
I'll take a look at it today i believe i have a solution for this

**Comment by anandwana001** [comment]
Thanks, this will be helpful.

**Comment by khushal1707** [comment]
@anandwana001 So the problem which I understood is that grouping items doesn't invoke DiffUtil, whereas on dragging and dropping it refreshes and shows the the grouped item. I haven't found any function in DiffUtil to solve this issue but there is a workaround which invoke DiffUtil to refresh list i.e. to call function `onItemMoved()` in `DragAndDropSortInteractionViewModel` twice such that its effect on the list got nullify( this is just the workaround and doesn't completely solves the problem ). So should I implements these workarounds or try something else?

**Comment by anandwana001** [comment]
let's have. a draft pr, which solves this issue and will discuss more on this. Also, keeping @aggarwalpulkit596 in loop for drag and drop feature.

**Comment by khushal1707** [comment]
Sure!

**Comment by aggarwalpulkit596** [comment]
OnItemMoved won't work correctly if i remember what i did while trying to fix this.

**Comment by khushal1707** [comment]
> OnItemMoved won't work correctly if i remember what i did while trying to fix this.

Yes, it didn't refresh when there is only one item on list and also cashes app sometimes.
Is there any other way to invoke DiffUtil when grouping occurs @aggarwalpulkit596 ?

**Comment by aggarwalpulkit596** [comment]
I'd have to try it out myself first it's been quite some time 😅

**Comment by khushal1707** [comment]
> I'd have to try it out myself first it's been quite some time 😅

Okay, till then I also try to figure out something and if something works I'll make a draft PR and notify.

**Comment by khushal1707** [comment]
Hi @anandwana001 @aggarwalpulkit596, can we use `notifyDataSetChanged()` for nested recyclerview of `DragAndDropSortInteraction` only after merging in `DragAndDropSortInteractionViewModel` and use Diffing in main and other recyclerviews. Merging works fine with this approach.

**Comment by khushal1707** [comment]
@anandwana001 @aggarwalpulkit596 PTAL: https://github.com/oppia/oppia-android/pull/2666
Haven't added tests yet, will be adding as soon as issue with diffing resolves.
Thanks!

**Comment by BenHenning** [comment]
We discussed this again today as it relates to #4839. I've added significant context to the main issue body above to explain what we've tried, why it (probably) didn't work, and how we can (probably) approach this in a way that should work well.

**Comment by BenHenning** [comment]
@subhajitxyz are you still interested in taking this issue on (please make sure to look through the main issue body above), as discussed in the CLaM meeting today?

**Comment by subhajitxyz** [comment]
Yes , I am working on it.

**Comment by subhajitxyz** [comment]
HI @BenHenning , I was trying to implement for State Items RecyclerView by following your approach.
I understand most it. But I am struggling with that how to introduce unique Id for viewmodels.

Currently our stateitem list working like that --
we compute list of StateItem viewmodels from ephemeralstate. So every time user do some action like submitting answer or hint unlock. we compute from ephemeralstate.  So it becomes the source of the viewmodels.

That meas we need to store these ids in ephemeral state or we have to find a way so that viewmodels get the same id from old list.

(Also we can not depend on existing viewmodel properties as unique id. because viewmodels (like feedbackViewmodel) they can appear multiple time and have same properties.)

If I am missing something or misunderstood your approach. Could you please clarify?

**Comment by BenHenning** [comment]
@subhajitxyz I think you're hitting on why this is a challenging thing to solve, and why it hasn't been done yet. Essentially all view models need to be able to provide a unique ID, and as you pointed out this can't trivially use the content. However one thing that might work to mitigate the specific case you mentioned (feedback) would be to include the response index. We never go back and change old items in the list so indexes are reasonable indicators for unique IDs (only for state fragment--this doesn't apply generally and this issue is meant to solve the problem generally for all app recycler views).
